libtracepoint-control-cpp/include/tracepoint/TracingSession.h

```diff
@@ -0,0 +1,709 @@
+        - activeCount: On success, receives the number of buffers that meet the
+          wakeup condition, or 0 if wait ended due to a timeout or a signal.
+
+        Returns EPERM if the session is not realtime.
+
+        Returns EPERM if the session is inactive. After construction and after
+        Clear(), the session will be inactive until a tracepoint is added.
+
+        Note that wakeup conditions are evaluated per-buffer. For example, if 3
+        buffers each contain 32760 bytes of pending data, none of them would
+        trigger a WakeupWatermark(32768) condition.
+        */
+        _Success_(return == 0) int
+        WaitForWakeup(
+            timespec const* timeout = nullptr,
+            sigset_t const* sigmask = nullptr,
+            _Out_opt_ int* pActiveCount = nullptr) noexcept;
+
+        /*
+        Advanced scenarios: Returns the file descriptors used for the buffers
+        of the session. The returned file descriptors may be used for poll()
+        but should not be read-from, written-to, closed, etc. This may be
+        useful if you want to use a single thread to poll for events from
+        multiple sessions or to poll for both events and some other condition.
+
+        Returns EPERM if the session is inactive. After construction and after
+        Clear(), the session will be inactive until a tracepoint is added.
+
+        Most users should use WaitForWakeup() instead of GetBufferFiles().
+        */
+        _Success_(return == 0) int
+        GetBufferFiles(
+            _Out_writes_(BufferCount()) int* pBufferFiles) const noexcept;
+
+        /*
+        For each PERF_RECORD_SAMPLE record in the session's buffers, invoke:
+        
+            int error = eventInfoCallback(eventInfo, args...);
+
+        - eventInfoCallback: Callable object (e.g. a function pointer or a lambda)
+          to invoke for each event.
+          
+          This callback should return an int (0 for success, errno for error). If
+          eventInfoCallback returns a nonzero value then EnumerateEventsUnordered
+          will immediately stop and return the specified error value.
+          
+          This callback should take a PerfSampleEventInfo const& as its first
+          parameter.
+          
+          The args... (if any) are from the args... of the call to
+          EnumerateEventsUnordered(eventInfoCallback, args...).
+
+        - args...: optional additional parameters to be passed to eventInfoCallback.
+
+        Returns: int error code (errno), or 0 for success.
+
+        For efficiency, events will be provided in a natural enumeration order.
+        This is usually not the same as event timestamp order, so you may need to
+        sort the events after receiving them.
+
+        Note that the eventInfo provided to eventInfoCallback will contain pointers
+        into the trace buffers. The pointers will become invalidated after
```

Comment:
This kind of lifetime is really hard to express in Rust.

I know this C++ isn't suitable for FFI, but just as a general pattern if other languages are going to follow this callback pattern it may be harder to use than, say, an iterator or async iterator.

Reply:
I don't think other languages are going to follow this callback pattern, and if they do, it's just invoking a normal callback or closure of the caller's choice. I'm not sure what makes it hard to express.

I do agree that an iterator would be easier to work with, but it would also create other issues. For example, you can only have one enumeration going at a time, it needs to clean up when it's done, the class is not inherently thread-safe, etc.

Reply:
I think I might understand what's bugging you. This API is kind of like ProcessTrace(file), but is not at all like ProcessTrace(realtime). There is no blocking involved in this API, no opportunity for async. The idea of this API is to get in, report the events, and get out, as quickly as possible. Time spent in this API is supposed to be minimized and is definitely non-blocking.

In the case of realtime sessions, the blocking is done outside of this function.